src/Main.elm

```diff
@@ -0,0 +1,375 @@
+
+
+type GameState
+    = Started
+    | Finished (Maybe Player)
+
+
+
+-- The `Board` type is an alias for a dictionary mapping a position (x,y) to a player (X or O).
+
+
+type alias Board =
+    Dict ( Int, Int ) Player
+
+
+
+-- The initial model, when the game starts.
+
+
+initialModel : Model
+initialModel =
+    { board = Dict.empty -- An empty board.
+    , currentPlayer = X -- Player X starts.
+    , gameState = Started -- The game is just starting.
+    , boardSize = 3 -- A 3x3 board.
+    , toggleModal = True -- Show the initial play modal.
+    }
+
+
+type Msg
+    = SquareClick ( Int, Int ) -- A square on the board was clicked.
+    | AddCell -- Increase the board size by 1.
+    | DecCell -- Decrease the board size by 1.
+    | PlayAgain -- Start a new game.
+    | ToggleModal -- Hide the modal to change the board size.
+
+
+
+-- This function change the player after each turn
+
+
+changePlayer : Player -> Player
+changePlayer player =
+    case player of
+        X ->
+            O
+
+        O ->
+            X
+
+
+update : Msg -> Model -> Model
+update msg model =
+    case msg of
+        SquareClick ( x, y ) ->
+            -- If the game has ended, do nothing.
+            if model.gameState /= Started then
+                model
+
+            else
+                -- Insert the new move into the board.
+                let
+                    newBoard =
+                        Dict.insert ( x, y ) model.currentPlayer model.board
+                in
+                -- If the move was valid, switch players and check for a winner or a draw.
+                if isEmpty model.board ( x, y ) then
+                    { model
+                        | currentPlayer = changePlayer model.currentPlayer
+                        , board = newBoard
+                        , gameState =
+                            if Dict.size model.board < (model.boardSize * model.boardSize - 1) && checkWinner newBoard model == Nothing then
+                                Started
+
+                            else
+                                Finished (checkWinner newBoard model)
+                    }
+
+                else
+                    -- If the move was invalid, do nothing.
+                    model
+
+        PlayAgain ->
+            -- Reset the game to the initial state.
+            initialModel
+
+        AddCell ->
+            -- Increase the board size by 1.
+            let
+                newBoardSize =
+                    if model.boardSize > 4 then
+                        -- I have limited here to 5 squares at most, but this can work for any number
+                        5
+
+                    else
+                        model.boardSize + 1
+            in
+            { model
+                | boardSize = newBoardSize
+                , board = Dict.empty
+                , currentPlayer = X
+                , gameState = Started
```

Comment:
why are we changing this here? This is work of PlayAgain, isn't it? 🤔  (same for DecCell)